File: src/core/alerts.cpp
#include "core/alerts.h"

#include <string_view>
#include <unordered_map>

namespace cusage {

std::vector<std::pair<std::string, UsageWindow>> enumerate_windows(
    const UsageSnapshot& snapshot) {
  std::vector<std::pair<std::string, UsageWindow>> windows;
  if (snapshot.five_hour_) windows.emplace_back("5-hour", *snapshot.five_hour_);
  if (snapshot.seven_day_) windows.emplace_back("weekly", *snapshot.seven_day_);
  for (const auto& [model, window] : snapshot.model_weekly_) {
    windows.emplace_back("weekly:" + model, window);
  }
  return windows;
}

std::vector<Alert> compute_alerts(const UsageSnapshot& previous,
                                  const UsageSnapshot& current,
                                  const NotifyConfig& config) {
  std::vector<Alert> alerts;
  std::unordered_map<std::string, UsageWindow> previous_by_name;
  for (auto& [name, window] : enumerate_windows(previous)) {
    previous_by_name.emplace(name, window);
  }

  // A freed 5-hour window is useless while the weekly cap is exhausted, so a
  // 5-hour reset is suppressed when the current weekly window is at 100%.
  const bool weekly_consumed =
      current.seven_day_ && current.seven_day_->percent() >= 100;

  for (const auto& [name, window] : enumerate_windows(current)) {
    const auto it = previous_by_name.find(name);
    if (it == previous_by_name.end()) continue;  // New window: no baseline.

    const int before = it->second.percent();
    const int now = window.percent();
    const bool is_five_hour = (name == "5-hour");

    if (config.notify_on_drop && now < before) {
      // The 5-hour window only matters near the limit: alarm only when it was
      // at/above the floor. Weekly windows alarm on any decrease. And a 5-hour
      // reset is pointless while the weekly cap is fully consumed.
      if ((!is_five_hour || before >= config.five_hour_drop_floor) &&
          !(is_five_hour && weekly_consumed)) {
        alerts.push_back({AlertKind::window_dropped, name, now,
                          name + " freed: " + std::to_string(before) + "% -> " +
                              std::to_string(now) + "%"});
      }
      continue;  // A drop subsumes the rising-threshold check.
    }

    // Rising-edge alert applies to weekly windows only (the 5-hour window is
    // governed solely by the drop rule above).
    if (!is_five_hour && config.weekly_threshold > 0 &&
        before < config.weekly_threshold && now >= config.weekly_threshold) {
      alerts.push_back({AlertKind::threshold_crossed, name, now,
                        name + " at " + std::to_string(now) + "% (>= " +
                            std::to_string(config.weekly_threshold) + "%)"});
    }
  }
  return alerts;
}

}  // namespace cusage

File: android/app/src/main/java/no/automasjon/claudeusage/Alerts.kt
package no.automasjon.claudeusage

enum class AlertKind { DROPPED, THRESHOLD }

data class Alert(val kind: AlertKind, val window: String, val percent: Int, val message: String)

data class AlertConfig(
    val notifyOnDrop: Boolean = true,
    val weeklyThreshold: Int = 80,       // Weekly rising-edge alert %; 0 disables.
    val fiveHourDropFloor: Int = 95,     // 5-hour drop alarms only when prev% >= this.
)

// Faithful port of the C++ compute_alerts. Empty previous → no alerts.
// Per-window rules: weekly windows alarm on any decrease and on a rising
// weekly-threshold crossing; the 5-hour window alarms on a decrease only when its
// previous value was at/above fiveHourDropFloor (and has no rising alert).
// A 5-hour reset is suppressed while the weekly window is fully consumed (100%),
// since a freed 5-hour window is useless when the weekly cap is exhausted.
fun computeAlerts(
    previous: List<NamedWindow>,
    current: List<NamedWindow>,
    config: AlertConfig,
): List<Alert> {
  val previousByName = previous.associateBy { it.name }
  val weeklyConsumed = (current.firstOrNull { it.name == "weekly" }?.percent ?: 0) >= 100
  val alerts = ArrayList<Alert>()
  for (window in current) {
    val before = previousByName[window.name] ?: continue
    val beforePct = before.percent
    val nowPct = window.percent
    val isFiveHour = window.name == "5-hour"

    if (config.notifyOnDrop && nowPct < beforePct) {
      if ((!isFiveHour || beforePct >= config.fiveHourDropFloor) &&
          !(isFiveHour && weeklyConsumed)) {
        alerts.add(Alert(AlertKind.DROPPED, window.name, nowPct,
            "${window.name} freed: $beforePct% -> $nowPct%"))
      }
      continue
    }
    if (!isFiveHour && config.weeklyThreshold > 0 &&
        beforePct < config.weeklyThreshold && nowPct >= config.weeklyThreshold) {
      alerts.add(Alert(AlertKind.THRESHOLD, window.name, nowPct,
          "${window.name} at $nowPct% (>= ${config.weeklyThreshold}%)"))
    }
  }
  return alerts
}
